src/antsibull/versions.py

```diff
@@ -142,3 +152,176 @@ def load_constraints_if_exists(filename: str) -> dict[str, SemVerSpec]:
+    ansible_core_versions: Sequence[str],
+    ansible_core_version: PypiVer,
+    pre: bool = False,
+) -> PypiVer | None:
+    """
+    Retrieve the latest ansible-core bugfix release's version for the given ansible-core version.
+
+    :arg ansible_core_versions: A list of ansible-core versions.
+    :arg ansible_core_version: A ansible-core version.
+    """
+    versions = [PypiVer(v) for v in ansible_core_versions]
+    next_version = PypiVer(
+        f"{ansible_core_version.major}.{ansible_core_version.minor + 1}a"
+    )
+    newer_versions = [
+        version
+        for version in versions
+        if ansible_core_version <= version < next_version
+        and (pre or not version.is_prerelease)
+    ]
+    return max(newer_versions) if newer_versions else None
+
+
+def get_latest_collection_version(
+    versions: Sequence[str],
+    collection: str,
+    version_spec: str | None = None,
+    pre: bool = False,
+    prefer_pre: bool = False,
+    constraint: SemVerSpec | None = None,
+) -> SemVer:
+    """
+    Get the latest version of a collection that matches a specification.
+
+    :arg versions: Sequence of collection versions
+    :arg collection: Namespace.collection identifying a collection.
+    :arg version_spec: Optional string specifying the allowable versions.
+    :kwarg pre: If ``True``, allow prereleases (versions which have the form X.Y.Z.SOMETHING).
+        This is **not** for excluding 0.Y.Z versions.  Non-pre-releases are still
+        preferred over pre-releases, except if ``prefer_pre=True`` (for instance, with
+        ``version_spec='>2.0.0-a1,<3.0.0'`` and ``pre=True``, if the available versions
+        are 2.0.0-a1 and 2.0.0-a2, then 2.0.0-a2 will be returned.  If the available
+        versions are 2.0.0 and 2.1.0-b2, 2.0.0 will be returned since non-pre-releases
+        are preferred.) The default is ``False``.
+    :kwarg prefer_pre: If ``True``, prefer newer pre-releases over stable releases.
+    :kwarg constraint: If provided, only consider versions that match this specification.
+    :returns: :obj:`semantic_version.Version` of the latest collection version that satisfied
+        the specification.
+
+    .. seealso:: For the format of the version_spec, see the documentation
+        of :obj:`semantic_version.SimpleSpec`
+    """
+    sem_versions = [SemVer(v) for v in versions]
+    sem_versions.sort(reverse=True)
+
+    if version_spec:
+        spec = SemVerSpec(version_spec)
+        sem_versions = [v for v in sem_versions if v in spec]
+
+    if constraint:
+        # Ignore all versions that do not match the constraints
+        sem_versions = [v for v in sem_versions if v in constraint]
+
+    prereleases = []
+    for version in sem_versions:
+        # If this is a pre-release, first check if there's a non-pre-release that
+        # will satisfy the version_spec.
+        if version.prerelease:
+            # If we prefer prereleases, take this one.
+            if pre and prefer_pre:
```

Comment:
If `prefer_pre` is True and `pre` is False, I think the latest pre-release should still be returned. 

```suggestion
            if prefer_pre:
```

Reply:
I disagree, if `pre` is `False` prereleases should never be returned.

Reply:
Okay, if you feel strongly about that, fine with me.